tutorials/preprocessing/plot_55_setting_eeg_reference.py

```diff
@@ -0,0 +1,196 @@
+# use average of mastoid channels as reference
+# raw.set_eeg_reference(ref_channels=['M1', 'M2'])
+
+###############################################################################
+# If a scalp electrode was used as reference but was not saved alongside the
+# raw data (reference channels often aren't), you may wish to add it back to
+# the dataset before re-referencing. For example, if your EEG system recorded
+# with channel ``EEG 999`` as the reference but did not include ``EEG 999`` in
+# the data file, using :meth:`~mne.io.Raw.set_eeg_reference` to set (say)
+# ``EEG 050`` as the new reference will then subtract out ``EEG 050``'s signal
+# *without restoring the signal at* ``EEG 999``). In this situation, you can
+# add back ``EEG 999`` as a flat channel prior to re-referencing using
+# :func:`~mne.add_reference_channels`. Here's how the data looks in its
+# original state:
+
+raw.plot()
+
+###############################################################################
+# By default, :func:`~mne.add_reference_channels` returns a copy, so we can go
+# back to our original ``raw`` object later. If you wanted to alter the
+# existing :class:`~mne.io.Raw` object in-place you could specify
+# ``copy=False``.
+
+# add new reference channel (all zero)
+raw_new_ref = mne.add_reference_channels(raw, ref_channels=['EEG 999'])
+raw_new_ref.plot()
+
+###############################################################################
+# .. KEEP THESE BLOCKS SEPARATE SO FIGURES ARE BIG ENOUGH TO READ
+
+# set reference to `EEG 050`
+raw_new_ref.set_eeg_reference(ref_channels=['EEG 050'])
+raw_new_ref.plot()
+
+###############################################################################
+# Notice that ``EEG 050`` is now flat, while ``EEG 999`` has a non-zero signal,
+# and that ``EEG 053`` (which is marked as "bad" in ``raw.info['bads']``) is
+# not affected by the re-referencing.
+#
+#
+# Setting average reference
+# ^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# To set a "virtual reference" that is the average of all channels, you can use
+# :meth:`~mne.io.Raw.set_eeg_reference` with ``ref_channels='average'``. Just
+# as above, this will not affect any channels marked as "bad", nor will it
+# include bad channels when computing the average. However, it does modify the
+# :class:`~mne.io.Raw` object in-place, so we'll make a copy first so we can
+# still go back to the unmodified :class:`~mne.io.Raw` object later:
+
+# sphinx_gallery_thumbnail_number = 4
+# use the average of all channels as reference
+raw_avg_ref = raw.copy().set_eeg_reference(ref_channels='average')
+raw_avg_ref.plot()
+
+###############################################################################
+# Creating the average reference as a projector
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# If using an average reference, it is possible to create the reference as a
+# :term:`projector` rather than subtracting the reference from the data
+# immediately, by specifying ``projection=True``:
```

Comment:
I think this is a bit underdescribed. What's the point in doing this? When might you want to do this?